src/Symfony/Component/Validator/Test/ConstraintValidatorTestCase.php

```diff
@@ -391,6 +391,17 @@ class AssertingContextualValidator implements ContextualValidatorInterface
     private $validateCalls = -1;
     private $expectedValidate = [];
 
+    public function __destruct()
+    {
+        if ($this->expectedAtPath) {
+            throw new ExpectationFailedException('Some expected validation calls for paths were not done.');
+        }
```

Comment:
@fancyweb @nicolas-grekas 

I really understand the reason for this fix but I've noticed something when I've upgraded from 4.4.36 to 4.4.44

My validator does a validation in context but it does not call `atPath`. I am passing the same instance but with a group based on a object's property.
```php
// public function validate($value, Constraint $constraint)
$this
    ->context
    ->getValidator()
    ->inContext($this->context)
    ->validate($value, null, [sprintf('entity_%s', $value->getType())])
;
```

When I do the test expectation, I pass a `null` value as `$propertyPath`. The `expectValidateAt` will register both the `expectedAtPath` and the `expectedValidate`.
```php
// public function testValidationInContextWithEntityType(): void
$entity = new Entity();
$entity->setType('foo');

$this->expectValidateAt(0, null, $entity, ['entity_foo']);
$this->validator->validate($entity, new EntityConstraint());
```

As `atPath` is not called in the validator's implementation, the `AssertingContextualValidator::doAtPath` is not called either, and therefore the `expectedAtPath` is not unset correctly and the exception is thrown.

---
**Is it mandatory to call `atPath` in a validation in context?**

If it is not, then I see some flaws in the ConstraintValidatorTestCase file. As both `expectValidateAt` and `expectValidateValueAt` forces me to pass a property path, the only valid value is a `null` in my scenario.

Of course, I could fix the tests in different ways, e.g. I can think of...
- Bypassing the exception with some ugly code to execute the `AssertingContextualValidator::doAtPath` manually:
```php
// This is still failing because the doAtPath evaluates the null value,
// from expectedAtPath, as false `!isset(null)` == true
$this->expectValidateAt(0, null, $entity, ['entity_foo']);
$this->validator->validate($entity, new EntityConstraint());
$this->context->getValidator()->inContext($this->context)->atPath(null);
// This works but it is not mirroring what the validator really does
$this->expectValidateAt(0, '', $entity, ['entity_foo']);
$this->validator->validate($entity, new EntityConstraint());
$this->context->getValidator()->inContext($this->context)->atPath('');
```
- Executing `atPath` in my validator with an empty string (I hope this is harmless and it will not produce any unexpected behavior)
```php
$this
    ->context
    ->getValidator()
    ->inContext($this->context)
    ->atPath('')
    ->validate($value, null, [sprintf('entity_%s', $value->getType())])
;
```

However, I wouldn't like to change the validator logic due to abstract test cases. Some advise?

Thanks in advance!